test/server/cards/02_SHD/units/PunishingOne.spec.ts

```diff
@@ -0,0 +1,27 @@
+describe('Punishing One', function() {
+    integration(function(contextRef) {
+        describe('Punishing One\'s ability', function() {
+            beforeEach(function () {
+                contextRef.setupTest({
+                    phase: 'action',
+                    player1: {
+                        hand: ['vanquish'],
+                        spaceArena: [{ card: 'punishing-one#dengars-jumpmaster', exhausted: true }],
+                    },
+                    player2: {
+                        groundArena: [{ card: 'wampa', upgrades: ['experience'] }],
+                    }
+                });
+            });
+
+            it('should be able to ready when an upgraded enemy unit is defeated', function () {
```

Comment:
A few more test cases would be good to confirm we have the conditions set up right. Note that unfortunately, due to the fact that Jasmine tests are really heavy and we have finite Github Actions time in the free tier, we have to kind of chain tests together in single cases. See the Rukh tests for the typical pattern we use for this to keep it clean.

Below are the tests it would be good to extend this case with. Since the ability limit is once per round, you can use `context.moveToNextActionPhase()` in between to reset the limit.

- Does not ready when a non-upgraded enemy unit is defeated
- Does not ready when an upgraded friendly unit is defeated
- Test the ability limit by defeating a second upgraded enemy unit in the same phase after the ability has been triggered once already
- An interesting test would be to have Punishing One defeat an upgraded unit in combat, use its ability on itself, then confirm it is ready